### PR TITLE
Dolfinx 0.9.0

### LINKED ISSUE #50: Usage of `fem.Expression` compared to `fem.Function.interpolate`
Hi @srosenbu 

I just noticed that something like
```python
q = fem.Function(Q) # Q is the scalar quadrature space
q.interpolate(lambda x: x[0], cells=cells)
```
will fill the entries according to cells in `q`  and leave other entries being zero.

Using `fem.Expression` on the other hand, only the nonzero values would be returned or the user would need to provide an array of the correct size to store the result.

That being said, I don't think using `interpolate` for `_del_grad_u` (see [code](https://github.com/BAMresearch/fenics-constitutive/blob/main/src/fenics_constitutive/solver.py#L241)) will work. What's more, the cell indices are also global indices, so the function on the subspace would probably complain if it encounters indices out of range, and you would evaluate the wrong cells.

We could, however, use `strain.interpolate` with `strain` as global function and call `interpolate` with cells for each material (so $n$ times if there are $n$ different materials).
But, then we would need to slice `strain.x.array` when passing the values to the different material routines.

Best,
Philipp

### PR DESCRIPTION
**Pull request by srosenbu**
This PR will add

* [x] dolfinx v0.9.0 support
* [x] MPI functionality
* [x] Rigorous testing of MPI functionality -- CI test that compares a plasticity computation on `COMM_SELF` to one on `COMM_WORLD`
* [x] Translate material model tests
* ~[] Rewriting of material model tests to be compatible with MPI~We don't need that for one-element tests
*  [] Not needed, since our `IncrSmallStrainProblem` does not create petsc objects and vectors implement their own `__del__` ~check if we need to reimplement `__del__` (https://github.com/FEniCS/dolfinx/blob/a547191a3c2765acd12aeadd0a3a66c0b99e83a9/python/dolfinx/nls/petsc.py#L44)~
* [] Create new issue~Try to use the subspace stresses and tangents to build the ufl forms. This would reduce memory significantly. Then only the displacements would live on the whole mesh and everything else on the submeshes (see https://github.com/FEniCS/dolfinx/blob/bde06f598731cdcd53d36275c92dd11cc5feaba1/python/test/unit/fem/test_assemble_submesh.py )~

Resolves #50, #47

### CODE REVIEW

**Comment by srosenbu** [comment]
@pdiercks , Do you have any good ideas on how we can test the MPI functionality?
Ideally we should test if the solution fields and quadrature fields are identical for some Problem that has been solved with a different number of MPI processes. 

Can we try to write a pytest case where we solve a distributed Problem with `MPI.COMM_WORLD` and on one of the ranks the same problem with `MPI.COMM_SELF`? That way we could produce an output within one test case without writing any results to the disk. But I don't know if we can mix the comms. For example would the following work?:

```python3
mesh_distr = df.mesh.create_unit_cube(MPI.COMM_WORLD,10,10,10)
mesh_local = df.mesh.create_unit_cube(MPI.COMM_SELF,10,10,10)

space_distr = df.fem.functionspace(mesh_distr, ("Lagrange",1))
space_local = df.fem.functionspace(mesh_local, ("Lagrange",1))

u_distr = df.fem.Function(space_distr)
u_local = df.fem.Function(space_local)

diff = u_distr - u_local
```

**Comment by srosenbu** [comment]
I tried it. Does not work with different communicators. I will try https://gitlab.com/dglaeser/fieldcompare

> @pdiercks , Do you have any good ideas on how we can test the MPI functionality? Ideally we should test if the solution fields and quadrature fields are identical for some Problem that has been solved with a different number of MPI processes.
> 
> Can we try to write a pytest case where we solve a distributed Problem with `MPI.COMM_WORLD` and on one of the ranks the same problem with `MPI.COMM_SELF`? That way we could produce an output within one test case without writing any results to the disk. But I don't know if we can mix the comms. For example would the following work?:
> 
> ```python
> mesh_distr = df.mesh.create_unit_cube(MPI.COMM_WORLD,10,10,10)
> mesh_local = df.mesh.create_unit_cube(MPI.COMM_SELF,10,10,10)
> 
> space_distr = df.fem.functionspace(mesh_distr, ("Lagrange",1))
> space_local = df.fem.functionspace(mesh_local, ("Lagrange",1))
> 
> u_distr = df.fem.Function(space_distr)
> u_local = df.fem.Function(space_local)
> 
> diff = u_distr - u_local
> ```

File: src/fenics_constitutive/error_estimation.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import ufl
+from mpi4py import MPI
+
+
+def norm(f, dx, comm=MPI.COMM_WORLD, type="l2"):
+    match type:
+        case "l2":
+            norm_squared = df.fem.assemble_scalar(df.fem.form(ufl.inner(f, f) * dx))
+            return np.sqrt(comm.allreduce(norm_squared, op=MPI.SUM))
+        case "inf":
```
[the remark this answers is not in the document]
Comment (srosenbu) [reply src/fenics_constitutive/error_estimation.py:14]: This is the responsibility of the caller, not the function

File: src/fenics_constitutive/models/mises_plasticity_isotropic_hardening.py [hunk 2]
```diff
@@ -103,10 +103,12 @@ def evaluate(
             # elastic - plastic step
             if phitr > 0:
                 # initialization
-                gamma = 0
+                gamma_0 = 1
```
[the remark this answers is not in the document]
Comment (srosenbu) [reply src/fenics_constitutive/models/mises_plasticity_isotropic_hardening.py:106]: This ois done in order for the first residual to be 1 instead of zero. This is done to not pass the first convergence check

File: README.md [hunk 3]
```diff
@@ -47,7 +48,7 @@ youngs_modulus = 42.0
 poissons_ratio = 0.3
 
 mesh = df.mesh.create_unit_cube(MPI.COMM_WORLD, 2, 2, 2)
-V = df.fem.VectorFunctionSpace(mesh, ("CG", 1))
+V = df.fem.functionspace(mesh, ("CG", 2, (2,)))
```
Comment (aradermacher) [inline README.md:51]: You changed the order of the function space from the old to the new version, right?
Just for my understanding, what's the second 2?

File: src/fenics_constitutive/models/mises_plasticity_isotropic_hardening.py
Comment (aradermacher) [inline src/fenics_constitutive/models/mises_plasticity_isotropic_hardening.py:1]: Looks fine for me, did you ask Saif to check?

File: src/fenics_constitutive/solver.py [hunk 4]
```diff
@@ -30,29 +32,16 @@ def build_history(
...
-                    q_degree,
-                    quad_scheme="default",
-                    shape=value,
-                )
-        history_space = df.fem.FunctionSpace(mesh, Qh)
+        value_shape = (value,) if isinstance(value, int) else value
+        Q = basix.ufl.quadrature_element(
+            mesh.topology.cell_name(), value_shape=value_shape, degree=q_degree
+        )
+        history_space = df.fem.functionspace(mesh, Q)
```
Comment (aradermacher) [inline src/fenics_constitutive/solver.py:39]: That's much less code for the same stuff. Nice

File: tests/models/test_plasticity.py
Comment (aradermacher) [inline tests/models/test_plasticity.py:1]: same as before. Let Saif have a brief check

File: tests/models/test_viscoelasticity.py [hunk 5]
```diff
@@ -291,7 +291,7 @@ def test_kelvin_vs_maxwell():
     """1D test with uniaxial stress, compare Kelvin and Maxwell model."""
 
     mesh = df.mesh.create_unit_interval(MPI.COMM_WORLD, 2)
-    V = df.fem.FunctionSpace(mesh, ("CG", 1))
+    V = df.fem.functionspace(mesh, ("CG", 1))
```
Comment (aradermacher) [inline tests/models/test_viscoelasticity.py:294]: Sometimes you have ("CG",1 (dim,)) Here the dim part is missing. Why? Is it default 1?

File: tests/test_solver_mpi.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    adios4dolfinx.write_function_on_input_mesh(
+        f"mpi_test_{comm_name}.bp",
+        u,
+        mode=adios4dolfinx.adios2_helpers.adios2.Mode.Write,
+    )
+    return f"mpi_test_{comm_name}.bp"
+
+
+@pytest.mark.mpi
+def test_mpi_solver():
```
Comment (aradermacher) [inline tests/test_solver_mpi.py:91]: and how much faster is the run with mpi?

**Review by aradermacher — APPROVED** [review]
Look's fine. Thanks for the work.

I didn't run the test at my machine. I trust github :)

File: src/fenics_constitutive/models/mises_plasticity_isotropic_hardening.py
Comment (saifr68) [reply src/fenics_constitutive/models/mises_plasticity_isotropic_hardening.py:1]: Thanks for updating the newton raphson convergence condition. Looks much better than before.

File: tests/models/test_plasticity.py
Comment (saifr68) [reply tests/models/test_plasticity.py:1]: Boundary conditions were incomplete in the previous version which have been corrected now. Thanks

**Review by saifr68 — COMMENTED** [review]
Nice work
